### ISSUE TITLE
Allow custom cache implementation

### ISSUE THREAD

**Issue by brstgt** [PHP-DI/PHP-DI#589]
Currently, the cache implementation is hard-wired. I'd like to replace the apcu cache with an in-memory cache.

Why?
- I don't want to roll out apc just for that
- I don't want a persistent cache
- I don't want a cache that shares a namespace among processes (if not configurable)
- I simply don't need it. We have  long running processes (amphp/aerys) and a in-mem cache is totally sufficient

Implementing a cache is super ease but I can't plug it in. To make it work I had to copy + adapt the whole container-builder :(

How a bout ContainerBuilder::setCache($cacheObj) or ::setCacheImplementation($className)?

**Comment by mnapoli** [comment]
To answer your first question:

> Why?

http://php-di.org/doc/performances.html#caching

> The cache relies on APCu directly because it is the only cache system that makes sense (very fast to write and read). Other caches are not good options, this is why PHP-DI does not use PSR-6 or PSR-16 for this cache.

Using Redis/Memcached for that would not make sense as it would be too slow.

In your use case (aerys) I understand an array cache (in-memory) could make sense though. Do you have performance issues if you disable APCu?

Re-introducing an abstraction for the cache could be doable, but I'd rather avoid that if I can because it's a pain to maintain, document and explain.

**Comment by brstgt** [comment]
Yes we did have immense performance penalties. An array-cache worked fine for us.
You don't need an abstraction IMHO. You could simply use existing interfaces. Actually it's only a `DefinitionSource` decorating all other sources, right?

**Comment by brstgt** [comment]
Otherwise add a simple option to use apcu or array-cache. An array cache is only a few lines different from apc cache

**Comment by mnapoli** [comment]
> Otherwise add a simple option to use apcu or array-cache

;) For specific feature requests I can work on commission if your company is willing to support the project. Else I will consider this issue (and the solutions) based on priority (number of people/projects impacted, interest/criticity for the project, etc.) and how much personal time I can spend :)

**Comment by brstgt** [comment]
I can offer a PR but I won't spend time on that if it's not clear that it will be accepted.
